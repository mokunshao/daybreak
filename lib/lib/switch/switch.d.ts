import React, { InputHTMLAttributes, ChangeEventHandler } from 'react';
import './switch.scss';
interface Props extends InputHTMLAttributes<HTMLInputElement> {
    checked: boolean;
    onChange: ChangeEventHandler<HTMLInputElement>;
}
export declare const Switch: React.FC<Props>;
export default Switch;
//# sourceMappingURL=switch.d.ts.map