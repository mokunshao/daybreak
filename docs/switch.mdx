---
title: Switch
---

import {useState} from 'react'
import { Playground, Props } from 'docz'
import Switch from '../src/switch/switch.tsx'

# Switch

```javascript
import { Switch } from 'daybreak';
```

## Standalone

<Playground>
  {()=>{
    const [value, setValue] = useState(true);
    return (
      <div>
        <Switch checked={value} onChange={(e) => setValue(e.target.checked)}></Switch>
      </div>
    );
  }}
</Playground>

## With Children

<Playground>
  {()=>{
    const [value, setValue] = useState(true);
    return (
      <div>
        <Switch checked={value} onChange={(e) => setValue(e.target.checked)}>Enable?</Switch>
      </div>
    );
  }}
</Playground>
