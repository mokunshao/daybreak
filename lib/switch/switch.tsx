import React, { InputHTMLAttributes, ChangeEventHandler } from 'react';
import { classes } from '../utils/classes';
import { joinedClass } from '../utils/joinedClass';

const baseClass = joinedClass('switch');

interface Props extends InputHTMLAttributes<HTMLInputElement> {
  checked: boolean;
  onChange: ChangeEventHandler<HTMLInputElement>;
}

export const Switch: React.FC<Props> = (props) => {
  const {
    checked, onChange, children, className, ...rest
  } = props;
  return (
    <label className={classes(baseClass(), className)}>
      <input type="checkbox" className={baseClass('input')} onChange={(e) => onChange(e)} checked={checked} {...rest} />
      <span className={baseClass('checkmark')} />
      {children}
    </label>
  );
};

export default Switch;
