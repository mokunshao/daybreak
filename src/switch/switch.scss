@import '../common.scss';

.daybreak-switch {
  display: inline-flex;
  align-items: center;
  cursor: pointer;

  &-input {
    visibility: hidden;
    height: 0;
    width: 0;
  }

  &-slider {
    display: inline-flex;
    width: 3em;
    height: 1.5em;
    background-color: $grey2;
    margin-right: 0.2em;
    margin-left: 0.2em;
    position: relative;
    box-sizing: border-box;
    transition: all 200ms;

    &::before {
      content: '';
      width: 1em;
      height: 1em;
      background-color: $white;
      position: absolute;
      left: 0.25em;
      top: 50%;
      transform: translateY(-50%);
    }
  }

  &-input:checked ~ &-slider {
    background-color: $black;
  }

  &-input:checked ~ &-slider::before {
    left: auto;
    right: 0.25em;
  }
}
