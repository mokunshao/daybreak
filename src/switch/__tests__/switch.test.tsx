import React, { useState } from 'react';
import { render } from '@testing-library/react';
import { Switch } from '../switch';

function Example() {
  const [value, setValue] = useState(true);
  return (
    <div>
      <Switch checked={value} onChange={(e) => setValue(e.target.checked)} />
    </div>
  );
}

describe('Switch', () => {
  it('exits', () => {
    expect(Switch).toBeTruthy();
    expect(<Example />).toBeTruthy();
  });

  it('match snapshot', () => {
    const { asFragment } = render(<Example />);
    expect(asFragment()).toMatchSnapshot();
  });
});
