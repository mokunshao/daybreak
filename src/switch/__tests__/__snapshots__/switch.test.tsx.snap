// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch match snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="daybreak-switch"
    >
      <input
        checked=""
        class="daybreak-switch-input"
        type="checkbox"
      />
      <span
        class="daybreak-switch-slider"
      />
    </label>
  </div>
</DocumentFragment>
`;
